src/java.base/share/classes/java/util/TreeMap.java

```diff
@@ -723,7 +724,7 @@ else if (cmp > 0)
      * remapping function modified this map
      */
     @Override
-    public V merge(K key, V value, BiFunction<? super V, ? super V, ? extends V> remappingFunction) {
+    public @PolyNull V merge(K key, V value, BiFunction<? super V, ? super V, ? extends @PolyNull V> remappingFunction) {
```

Comment:
```suggestion
    public @PolyNull V merge(K key, @NonNull V value, BiFunction<? super V, ? super V, ? extends @PolyNull V> remappingFunction) {
```

Would adding `@NonNull` to `value` make sense, to match the signature in `Map`?
(We would also need to add the import.)

Reply:
Thanks, I should really look at this method (and probably some other Java 8 additions in `Map`) more carefully.

I agree that `@NonNull` makes sense on `value`. (And the implementations in `Map`, `HashMap`, `TreeMap`, and presumably others perform a null check, as the doc's first sentence suggests but the "implementation requirements" do not.)

In fact, I think we should go a step further to make the `remappingFunction` be a `BiFunction<? super @NonNull V, ? super @NonNull V, ? extends @PolyNull V>`
- The first argument is the old value, which is passed only if it's non-null. (In the null case, `merge` inserts `value` and returns.)
- The second argument is `value`, which, as we just discussed, is non-null.

I've updated accordingly, and I'll tentatively do the same in https://github.com/jspecify/jdk/pull/14.